Move to aioresponses library to simplify aiohttp tests

NodeJSmith opened the pull request:
@abmantis Wanted to get your thoughts on this before I start working on the other tests files. The existing fixtures and test setup has always been very confusing for me and I think aioresponses can make everything a lot simpler. 

Let me know if you have any preferences or concerns on what I have so far and I can incorporate that before moving on to the other test files.

Note: This branch is off of feature/move_urls_to_backend_selector so you'll see those commits as well until that PR is approved, it's the `test_aircon.py` file that I'm focusing on here.

NodeJSmith commented:
@abmantis I went ahead and finished this up - the diff makes it look like a lot of changes but is all the same tests, just using aioresponses which lets us write them a lot cleaner and clearer. I also made sure to keep all the same checks, based on your feedback on the last PR. The one exception to that is this check in test_appliancesmanager:

```python
        if account_id is None:
            m.assert_any_call(URL(BACKEND_SELECTOR_MOCK.user_details_url), "GET")
        else:
            assert (
                "GET",
                URL(BACKEND_SELECTOR_MOCK.user_details_url),
            ) not in m.requests
```

This one just checks that we do call `user_details_url` if we don't have an account id, instead of explicitly checking if it's the first call. The reason for this is that aioresponses doesn't give us a good way (that I could find at least) to determine if this was called first, as it stores the calls in a dictionary that is keyed by method and url (e.g. ("POST", "url_value.url.com")).

If you'd like me to keep looking for a way to assert that it's first let me know, I'm positive that it _can_ be done, I just didn't want to put in a ton of time to do so if we are comfortable with this level of check.

The tests `test_auth_multiple_client_credentials` and `test_auth_bad_credentials` also got slimmed down some, because we were providing data to the mock that actually wasn't required for the test assertions, so I removed that data.

The other "big" changes were moving the json data to data files instead of putting it in python, which just feels cleaner, and adding washer and dryer data files - there are no tests using them yet, but I wanted to get them in there for future tests.

NodeJSmith commented:
Oh and I also updated the tests GH action to test again 3.11 and 3.12 since that is what we are supporting now

NodeJSmith commented:
@abmantis If you need me to break this up into smaller PRs I can. I wasn't sure how to do so with this, since it's all contained to tests, but if you give me some guidance on what would be easiest to digest I can work with that. I know it's a lot to look at in one go

abmantis commented:
Hello!
When I started adding the tests using aiohttp I did have some trouble and the result isn't great, I agree 😅
At the time I looked at aioresponses, but for some reason I were not able to use it, and ended up doing something similar to what HA does (or did at the time). 

I haven't had the chance to review this properly yet, but from a brief look I don't think there's a need to split it. Thanks for the concern!

abmantis commented on tests/test_aircon.py:
```diff
@@ -1,228 +1,144 @@
-    assert_appliance_setter_call(
-        appliance_http_client_mock,
+async def test_attributes():
+    backend_selector_mock = BackendSelectorMock()
```
I would prefer to keep this as a fixture and arg to the test, since that makes test dependencies explicit (same for the other tests).

abmantis commented on tests/test_aircon.py:
```diff
@@ -1,228 +1,144 @@
-        appliance_http_client_mock,
+async def test_attributes():
+    backend_selector_mock = BackendSelectorMock()
+    session = aiohttp.ClientSession()
```
can we also make this a fixture?

abmantis commented on tests/test_aircon.py:
```diff
@@ -1,228 +1,144 @@
+    session = aiohttp.ClientSession()
+    aircon = Aircon(
+        backend_selector_mock,
+        MagicMock(),
```
any reason to remove the auth_mock fixture? I know it currently does not do anything, but makes this a bit more explicit too. Also, if later we need to add some mock behavior to the auth part, we just have to update the fixture method.

abmantis commented on tests/test_aircon.py:
```diff
@@ -1,228 +1,144 @@
+)
+async def test_setters(method: Callable, argument: Any, expected: dict):
+    backend_selector_mock = BackendSelectorMock()
+    expected_template = {
```
```suggestion
    expected_payload = {
```

abmantis commented on tests/test_aircon.py:
```diff
@@ -1,228 +1,144 @@
+
+
+@pytest.mark.parametrize(
+    ["method", "argument", "expected"],
```
```suggestion
    ["method", "argument", "expected_json"],
```

abmantis replied:
btw, nice idea to use parametrized tests here!

abmantis commented on tests/test_aircon.py:
```diff
@@ -1,228 +1,144 @@
+    with patch("whirlpool.appliance.Appliance._create_headers", return_value={}):
+        session = aiohttp.ClientSession()
+        aircon = Aircon(backend_selector_mock, MagicMock(), SAID, session)
+        url = backend_selector_mock.appliance_command_url
```
move this down, near the line where it is first used

abmantis commented on tests/test_aircon.py:
```diff
@@ -1,228 +1,144 @@
+
+    with patch("whirlpool.appliance.Appliance._create_headers", return_value={}):
+        session = aiohttp.ClientSession()
+        aircon = Aircon(backend_selector_mock, MagicMock(), SAID, session)
```
move this down, near the line where it is first used

abmantis commented on tests/test_aircon.py:
```diff
@@ -1,228 +1,144 @@
+        }
+    }
+
+    call_kwargs = {
```
```suggestion
    post_request_call_kwargs = {
```

abmantis replied:
this could also be moved near its usage

abmantis commented on tests/test_aircon.py:
```diff
@@ -1,228 +1,144 @@
-    await aircon.set_display_on(True)
-    assert_appliance_setter_call(
-        appliance_http_client_mock,
+async def test_attributes():
```
```suggestion
@aioresponses()
async def test_request(mocked_aioresponses):
```
I think having this as a fixture instead of the context manager makes the tests a bit cleaner, and also reduce nesting

abmantis commented on tests/test_aircon.py:
```diff
@@ -1,228 +1,144 @@
+        (Aircon.set_display_on, True, {"Sys_DisplaySetBrightness": "4"}),
+    ],
+)
+async def test_setters(method: Callable, argument: Any, expected: dict):
```
```suggestion
async def test_setters(method: Callable, argument: Any, expected_json: dict):
```

abmantis reviewed (changes requested):
I have not reviewed everything yet, but have a couple of suggestions that probably also apply to the other tests.

Overall, I think this is a nice refactoring!

NodeJSmith commented:
@abmantis So I created some mocks to add to conftest.py and I noticed that `Auth` actually contains everything we need for all of our logic - it contains the backend selector, the auth credentials, and the session. We currently pass all of these to everything, but I think if we just accessed them from Auth it would work just as well.

Fixtures:
```python
@pytest.fixture
def aioresponses_mock():
    with aioresponses() as m:
        yield m


@pytest.fixture(params=[BackendSelectorMock, BackendSelectorMockMultipleCreds])
def backend_selector_mock(request):
    yield request.param()


@pytest_asyncio.fixture
async def auth_mock(backend_selector_mock):
    session = aiohttp.ClientSession()
    auth = Auth(backend_selector_mock, "email", "secretpass", session)
    yield auth
    await session.close()
```

Example - we previously had a backend selector mock we provided, plus the auth mock, plus the session. The new `auth_mock` fixture actually has all of these, so when we instantiate the Oven in the below test we can just use them all from the `Auth` object.

What are your thoughts on this? I don't want to pile up too many changes, but I'm not sure how to approach the tests. Does the below work for our tests or should I specifically separate out all of these pieces, even if it means more work for mocking the calls?

```python
async def test_attributes(auth_mock, aioresponses_mock):
    oven = Oven(auth_mock._backend_selector, auth_mock, SAID, auth_mock._session)
```

abmantis commented:
Regarding the Auth and mock thing, I would prefer to keep the fixtures instead of getting them from Auth, to keep responsibilities separated and also avoid accessing what should be internal stuff from Auth.

I am also wondering about an alternative: could/should we make Aircon/Oven/etc fixtures? That way, the tests only needs to receive that fixture as argument.

NodeJSmith commented:
In a general sense are you okay with all of these classes receiving instances of the same classes (backend selector, Auth, session) or should we add a review of the architecture to the backlog at some point? It seems like it could be a code smell.

I'm okay with keeping the fixtures separate, I'll break them back out.

Regarding the appliance fixtures, I think we can do that. I think they all use the same SAID, and we set the data using aioresponses, so that should be feasible. I'll take a look at it in the next day or two.

abmantis commented:
I actually like the fact that they are separated and each class receives the instances they need. I usually avoid classes that "agregate" things, as they tend to become dependency magnets. 
But I may be missing what you had in mind, so please do let me know if you have any thoughts on how to improve the arch.

NodeJSmith commented:
Whew, that took me longer than I'd planned.

I believe I implemented all of the changes you requested. I also took your advice and created mocks for the appliances, as well as mocks for the AppliancesManager and a fixture for the `aiohttp.ClientSession` with a `session` scope so we have the one for the whole duration of the tests and it closes itself. The tests look even cleaner now that they get everything through fixtures.

Two items I want to point out to you:

* I added the `with patch("whirlpool.appliance.Appliance._create_headers", return_value={}):` decorator in the appliance mocks, so that the headers were taken care of. I confirmed that all tests still pass as expected, so I think this is okay, but if you want me to do something differently here I can. My other idea was to overwrite the method with a patched method but I thought that decorating the whole fixture was probably cleaner until we have a reason to do something different.

* In test_appliancesmanager I added an autouse fixture, for the aioresponses calls that are required for all of the tests. Let me know if you have concerns with this - I wanted to make the tests look a bit cleaner but if you have concerns with an autouse fixture I can revert this. I also split out a lot of the assertions in this file into separate tests to make things more structured.

```python
@pytest.fixture(autouse=True)
def account_id_calls_fixture(aioresponses_mock, backend_selector_mock):
    """Mock the calls to get the user details and owned appliances.

    This is required by all of the tests in this file, so we use the autouse=True
    parameter to ensure that this fixture is always run.
    """
    aioresponses_mock.get(
        backend_selector_mock.user_details_url, payload={"accountId": ACCOUNT_ID}
    )
    aioresponses_mock.get(
        backend_selector_mock.get_owned_appliances_url(ACCOUNT_ID),
        payload={ACCOUNT_ID: owned_appliance_data},
    )
    aioresponses_mock.get(
        backend_selector_mock.shared_appliances_url, payload=shared_appliance_data
    )
    yield
```

abmantis commented on tests/test_appliancesmanager.py:
```diff
@@ -1,123 +1,148 @@
-        for k, v in required_headers.items():
-            assert call[3][k] == v
+with open(DATA_DIR / "owned_appliances.json") as f:
+    owned_appliance_data = json.load(f)
```
since this is only used in the fixture, we can move this inside it

abmantis commented on tests/test_appliancesmanager.py:
```diff
@@ -1,123 +1,148 @@
-    if forbidden_headers is not None:
-        for k in forbidden_headers:
-            assert k not in call[3]
+with open(DATA_DIR / "shared_appliances.json") as f:
```
same as above

abmantis commented on tests/test_appliancesmanager.py:
```diff
@@ -1,123 +1,148 @@
+    aioresponses_mock.get(
+        backend_selector_mock.shared_appliances_url, payload=shared_appliance_data
     )
+    yield
```
```suggestion
```
is that required?

abmantis commented on tests/test_appliancesmanager.py:
```diff
@@ -1,123 +1,148 @@
 
-    if account_id is not None:
-        auth._auth_dict["accountId"] = account_id
+async def test_uer_details_not_called_if_account_id_is_known(
```
```suggestion
async def test_user_details_not_requested_if_account_id_is_known(
```

abmantis commented on tests/test_appliancesmanager.py:
```diff
@@ -1,123 +1,148 @@
+    appliances_manager: AppliancesManager,
+    aioresponses_mock,
+):
+    auth_mock._auth_dict["accountId"] = ACCOUNT_ID
```
can we have this test without changing internal stuff of the class?
Maybe by:
- calling `fetch_appliances()` once
- checking the request was done
- resetting `aioresponses_mock.requests`
- calling `fetch_appliances()` again
- checking that the request was not done
 
If that works, we should probably rename the test to `test_user_details_requested_only_once` or something, and can remove the `test_user_details_called_if_account_id_not_known` test.

NodeJSmith replied:
Good idea, I've implemented this test and removed the other one

abmantis commented on tests/test_appliancesmanager.py:
```diff
@@ -1,123 +1,148 @@
+    shared_appliance_mock = get_mock_coro(None)
+    owned_appliance_mock = get_mock_coro(None)
+
+    appliances_manager._get_shared_appliances = shared_appliance_mock
+    appliances_manager._get_owned_appliances = owned_appliance_mock
```
Can we do this check using the http requests instead of altering internal behavior of the class? We could then probably also merge it `test_fetch_appliances_returns_owned_and_shared_appliances`

NodeJSmith replied:
Yep, that's a good idea, I've implemented it

abmantis commented on tests/test_appliancesmanager.py:
```diff
@@ -1,123 +1,148 @@
-    am = AppliancesManager(BACKEND_SELECTOR_MOCK, auth, http_client_mock.session)
-    am._get_shared_appliances = get_mock_coro(shared_response)
-    am._get_owned_appliances = get_mock_coro(owned_response)
+    appliances_manager._get_shared_appliances = get_mock_coro(shared_response)
```
Could also be done using the requests instead?

abmantis reviewed (changes requested):
Tests are indeed looking good! Left just some minor comments.
I haven't reviewed them all yet, hopefully I can review all the remaining before the weekend.

NodeJSmith commented:
@abmantis I moved `_get_account_id` from `AppliancesManager` to `Auth`, as it seems to make more sense there. The way it was previously, we could theoretically find that the account id is not in the `Auth`'s `_auth_dict` attribute, get the value using a call, and leave that empty in the `Auth` class, which doesn't make any sense.

Let me know if the changes look okay to you. I also removed the `test_fetch_appliances_returns_true_if_either_method_returns_true` test as it basically just tested boolean logic, which was silly.

I also updated the shared/owned tests to use the aioresponses assert method - this does require calling the private `_create_headers` method, but that's the only way to use the assert methods, and I think it's cleaner than the way I was doing it previously.

Oh and I fixed the cli command option, it was missing `await`s on the inputs.

Let me know if there are any concerns with the changes.

abmantis commented on tests/test_appliancesmanager.py:
```diff
@@ -45,51 +34,24 @@ def account_id_calls_fixture(aioresponses_mock, backend_selector_mock):
-        "GET",
-        URL(backend_selector_mock.user_details_url),
-    ) not in aioresponses_mock.requests
+    headers = appliances_manager._create_headers()
```
lets hard-code the headers here. that way we avoid accessing internal class stuff, and it also serves as a way to validate the headers and you can then remove the assert on line 54.

I don't really mind having some duplication or hardcoded strings on tests if they are used to validate the lib's behavior

abmantis commented on whirlpool/auth.py:
```diff
@@ -117,9 +117,28 @@ def is_access_token_valid(self):
         return self._auth_dict.get("access_token", None)
 
-    def get_account_id(self) -> str | None:
+    async def get_account_id(self) -> str | None | bool:
```
why not return None instead of False?

abmantis commented on tests/test_appliancesmanager.py:
```diff
@@ -99,13 +61,13 @@ async def test_get_shared_appliances_call_contains_wp_client_brand_header(
-    assert len(shared_call) == 1
-    shared_call_headers = shared_call[0][1]["headers"]
-    assert shared_call_headers["WP-CLIENT-BRAND"] == "DUMMY_BRAND"
+    headers = appliances_manager._create_headers(include_wp_brand_name=True)
```
same as above

abmantis commented on tests/test_appliancesmanager.py:
```diff
@@ -120,29 +82,20 @@ async def test_fetch_appliances_returns_owned_and_shared_appliances(
-    appliances_manager._get_shared_appliances = shared_appliance_mock
-    appliances_manager._get_owned_appliances = owned_appliance_mock
+    owned_headers = appliances_manager._create_headers()
+    shared_headers = appliances_manager._create_headers(include_wp_brand_name=True)
```
same as above

abmantis replied:
if we use hardcoded headers in this test, then do we need the `test_get_owned_appliances_call_does_not_contain_wp_client_brand_header` and `test_get_shared_appliances_call_contains_wp_client_brand_header` tests? maybe we can remove them

NodeJSmith replied:
Yep, good call, I removed those after hardcoding the headers

abmantis commented on tests/test_auth.py:
```diff
@@ -111,3 +114,35 @@ async def test_auth_bad_credentials(
+        backend_selector_mock.user_details_url, "GET", headers=headers
+    )
+
+    assert auth_mock._auth_dict["accountId"] == ACCOUNT_ID
```
```suggestion
```
you can reset `aioresponses_mock` and check that the call count for the URL is zero after the second call to `get_account_id()` instead of accessing the internal dict

abmantis commented on dryer_device_data_model.json:
what are these new json files for?

abmantis commented on tests/conftest.py:
```diff
@@ -1,48 +1,70 @@
 @pytest.fixture
-def backend_selector_mock():
-    return BackendSelectorMock()
+def auth_mock(backend_selector_mock, client_session_fixture):
```
```suggestion
pytest.fixture(name="auth")
def auth_fixture(backend_selector_mock, client_session_fixture):
```
seeing "auth_mock" in the tests where Auth is being tested was getting me confused, since it is not a mock

abmantis commented on tests/conftest.py:
```diff
@@ -1,48 +1,70 @@
-def auth_mock():
-    return MagicMock()
+@pytest.fixture(name="oven")
+def oven_mock(backend_selector_mock, auth_mock, client_session_fixture):
```
```suggestion
def oven_fixture(backend_selector_mock, auth_mock, client_session_fixture):
```

abmantis commented on tests/conftest.py:
```diff
@@ -1,48 +1,70 @@
+
+
+@pytest.fixture(name="washer_dryer")
+def washer_dryer_mock(backend_selector_mock, auth_mock, client_session_fixture):
```
```suggestion
def washer_dryer_fixture(backend_selector_mock, auth_mock, client_session_fixture):
```

abmantis commented on tests/conftest.py:
```diff
@@ -1,48 +1,70 @@
+
+
+@pytest.fixture(name="aircon")
+def aircon_mock(backend_selector_mock, auth_mock, client_session_fixture):
```
```suggestion
def aircon_fixture(backend_selector_mock, auth_mock, client_session_fixture):
```

abmantis commented on tests/conftest.py:
```diff
@@ -1,48 +1,70 @@
-    event_socket.stop = AsyncMock()
-    return event_socket
+@pytest.fixture(name="appliances_manager")
+def appliances_manager_mock(backend_selector_mock, auth_mock, client_session_fixture):
```
```suggestion
def appliances_manager_fixture(backend_selector_mock, auth_mock, client_session_fixture):
```

abmantis commented on tests/test_auth.py:
```diff
@@ -43,78 +39,110 @@ async def test_auth_success(http_client_mock: AiohttpClientMocker):
+    await auth_mock.do_auth(store=False)
+    assert auth_mock.is_access_token_valid()
+    assert auth_mock.get_said_list() == ["SAID1", "SAID2"]
+
```
```suggestion
    assert auth_mock.get_account_id() == ACCOUNT_ID
    
```

abmantis commented on tests/test_appliancesmanager.py:
```diff
@@ -1,123 +1,101 @@
-        for k in forbidden_headers:
-            assert k not in call[3]
+@pytest.fixture(autouse=True)
+def account_id_calls_fixture(aioresponses_mock, backend_selector_mock):
```
```suggestion
def common_http_calls_fixture(aioresponses_mock, backend_selector_mock):
```
I was going to suggest `user_details_and_appliances_calls_fixture`, but it was a bit too big...

abmantis commented on tests/test_appliancesmanager.py:
```diff
@@ -1,123 +1,101 @@
-    await http_client_mock.close_session()
+    aioresponses_mock.assert_called_with(
+        backend_selector_mock.shared_appliances_url, "GET", headers=shared_headers
+    )
```
can we also reset `aioresponses_mock`, call `fetch_appliances` again and assert that the call count to `aioresponses_mock` is zero?

NodeJSmith replied:
I didn't follow this comment, why would we do this? fetch_appliances will be calling the endpoints every time, it doesn't cache return appliances

abmantis replied:
Well, I don't remember why I suggested that. Now it does not seem to make sense at all :rofl:

NodeJSmith commented:
@abmantis Sorry to leave you hanging for a month, finally got around to implementing your requested changes, except for the last comment that I've left unresolved, I didn't follow what you were looking for on that one.

abmantis commented:
Thanks for getting back at this @NodeJSmith ! Everything LGTM. Are the 4 new json files needed, or can they be removed?

NodeJSmith commented:
I don't think the paths files or the data model files are required, I think those got added accidentally when I'd been messing with something else. I'll remove them

abmantis reviewed (approved):
Again, thanks a lot for this!
